compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -218,8 +217,12 @@ impl<'tcx> AbstractConst<'tcx> {
     }
 
     #[inline]
-    pub fn root(self) -> Node<'tcx> {
-        self.inner.last().copied().unwrap()
+    pub fn root(self, tcx: TyCtxt<'tcx>) -> Node<'tcx> {
+        let node = self.inner.last().copied().unwrap();
+        if let Node::Leaf(leaf) = node {
```

Comment:
I only realized this by looking at the code again, but things like `expr as *const T` have a parameter in the `Ty` of `Node::Cast`. So we should also call `subst` for these nodes :thinking: Wasn't able to trigger a bug with this as the I couldn't find a way to get `as SomethingWith<T>` into a generic const expr, but I still think we should "do the right thing" here.

Maybe make this an exhaustive match on `Node`, adding a comment like `// Can't directly contain generic params` to the branch for the kinds we can ignore

Reply:
nit: Just wondering if we could do something like `let node = self.inner.last().as_ref()` to avoid the copy ?

Reply:
we could, but considering that `Node` is fairly small I don't expect this to matter wrt performance